rust/core-lib/src/styles.rs

```diff
@@ -220,4 +240,152 @@ impl ThemeStyleMap {
         self.styles.push(style.clone());
         result
     }
+
+    /// Delete key and the corresponding dump file from the themes map.
+    pub(crate) fn remove_theme(&mut self, path: &Path) -> Option<String> {
+        validate_theme_file(path).ok()?;
+
+        let theme_name = path.file_stem().and_then(OsStr::to_str)?;
+        self.themes.themes.remove(theme_name);
+        self.state.remove(path);
+
+        let dump_p = self.get_dump_path(theme_name)?;
+        if dump_p.exists() {
+            let _ = fs::remove_file(dump_p);
+        }
+
+        Some(theme_name.to_string())
+    }
+
+    /// Load all themes inside the given directory.
+    pub(crate) fn load_theme_dir(&mut self) {
+        if let Some(themes_dir) = self.themes_dir.clone() {
+            match ThemeSet::discover_theme_paths(themes_dir) {
+                Ok(themes) => {
+                    self.caching_enabled = self.caching_enabled
+                        && self.init_cache_dir();
+
+                    for theme_p in themes.iter() {
+                        match self.try_load_from_dump(theme_p) {
+                            Some((k, v)) => {
+                                self.insert_to_map(k, v, theme_p);
+                            },
+                            None => {
+                                let _ = self.load_theme(theme_p);
+                            }
+                        }
+                    }
+                },
+                Err(e) => eprintln!("Error while loading themes dir: {:?}",e)
+            }
+        }
+    }
+
+    /// A wrapper around `from_dump_file`
+    /// to validate the state of dump file.
+    /// Invalidates if mod time of dump is less
+    /// than the original one.
+    fn try_load_from_dump(&self, theme_p: &Path)
+        -> Option<(String, Theme)>
+    {
+        if !self.caching_enabled { return None; }
+
+        let theme_name = theme_p.file_stem()
+            .and_then(OsStr::to_str)?;
+
+        let dump_p = self.get_dump_path(theme_name)?;
+
+        if !&dump_p.exists() { return None; }
+
+        let mod_t = fs::metadata(&dump_p)
+            .and_then(|md| md.modified()).ok()?;
+        let mod_t_org = fs::metadata(theme_p)
+            .and_then(|md| md.modified()).ok()?;
+
+        if mod_t >= mod_t_org {
```

Comment:
'org' usually means 'organization' whereas 'orig' is used for 'original' but this is now *really* in nit-picking territory and this section is pretty obvious otherwise. What is an interesting thing to consider, though, is how the case should be handled where a dump exists but the original file has been deleted.

Reply:
The iterator runs over the existing ".tmTheme" files in the "themes_dir" directory.
So, if a file is not present or is deleted it will never be loaded indirectly(via its dump) onto the editor's `theme` map.

Reply:
Never trust files to exist until you have actually read them 😉. I probably should have said 'where the original file is deleted between the creation of the iterator and the running of this fn on said file.' Either way, this code does the (arguably) reasonable thing of returning None in case of such situation. Consider leaving a note about it though.

Reply:
Something similar to hardlink.